pkg/bloomcompactor/bloomcompactor.go

```diff
@@ -463,92 +463,88 @@ func (c *Compactor) runCompact(ctx context.Context, logger log.Logger, job Job,
+	// It may create blocks with series outside of the fp range of the compactor. Cutting blocks will be addressed in a follow-up PR.
+	metasMatchingJob, blocksMatchingJob := matchingBlocks(metas, job)
+
+	localDst := createLocalDirName(c.cfg.WorkingDirectory, job)
+	blockOptions := v1.NewBlockOptions(bt.GetNGramLength(), bt.GetNGramSkip())
+
+	defer func() {
+		//clean up the bloom directory
+		if err := os.RemoveAll(localDst); err != nil {
+			level.Error(logger).Log("msg", "failed to remove block directory", "dir", localDst, "err", err)
+		}
+	}()
+	var resultingBlock bloomshipper.Block
+	if len(blocksMatchingJob) == 0 && len(metasMatchingJob) > 0 {
+		// There is no change to any blocks, no compaction needed
+		level.Info(logger).Log("msg", "No changes to tsdb, no compaction needed", "err", err)
+		return nil
+	} else if len(metasMatchingJob) == 0 {
+		// No matching existing blocks for this job, compact all series from scratch
 
-		blockOptions := v1.NewBlockOptions(bt.GetNGramLength(), bt.GetNGramSkip())
 		builder, err := NewPersistentBlockBuilder(localDst, blockOptions)
 		if err != nil {
 			level.Error(logger).Log("msg", "creating block builder", "err", err)
 			return err
 		}
 
 		fpRate := c.limits.BloomFalsePositiveRate(job.tenantID)
-		storedBlock, err := compactNewChunks(ctx, logger, job, fpRate, bt, storeClient.chunk, builder)
+		resultingBlock, err = compactNewChunks(ctx, logger, job, fpRate, bt, storeClient.chunk, builder)
 		if err != nil {
 			return level.Error(logger).Log("msg", "failed to compact new chunks", "err", err)
 		}
 
-		archivePath := filepath.Join(c.cfg.WorkingDirectory, uuid.New().String())
+	} else if len(blocksMatchingJob) > 0 {
+		// When already compacted metas exists, we need to merge all blocks with amending blooms with new series
 
-		blockToUpload, err := c.compressBloomBlock(storedBlock, archivePath, localDst, logger)
+		resultingBlock, err = mergeCompactChunks(ctx, logger, c.bloomShipperClient, storeClient, bt, job, blockOptions, blocksMatchingJob, c.cfg.WorkingDirectory, localDst)
 		if err != nil {
-			level.Error(logger).Log("msg", "putting blocks to storage", "err", err)
+			level.Error(logger).Log("msg", "failed to merge existing blocks with new chunks", "err", err)
 			return err
 		}
-		defer func() {
-			err = os.Remove(archivePath)
-			if err != nil {
-				level.Error(logger).Log("msg", "removing archive file", "err", err, "file", archivePath)
-			}
-		}()
+	}
 
-		// Do not change the signature of PutBlocks yet.
-		// Once block size is limited potentially, compactNewChunks will return multiple blocks, hence a list is appropriate.
-		storedBlocks, err := c.bloomShipperClient.PutBlocks(ctx, []bloomshipper.Block{blockToUpload})
+	archivePath := filepath.Join(c.cfg.WorkingDirectory, uuid.New().String())
+
+	blockToUpload, err := compressBloomBlock(resultingBlock, archivePath, localDst, logger)
```

Comment:
`resultingBlock.Data` needs to be closed

Reply:
it will be closed by the client https://github.com/grafana/loki/blob/5f4285963c5c4a796094e67afce51b2f67045094/pkg/storage/stores/shipper/bloomshipper/client.go#L231

Reply:
maybe it's responsibility of the compactor to close the reader. but then we need to remove closing the reader in the client

Reply:
It's used in multiple places, happy to do a follow up PR for this.

Reply:
> it will be closed by the client 

`resultingBlock` is not the object that is written using `PutBlocks`

Reply:
ah fair, thanks for catching. added.